1. 

(X_train=80%)
(X_test=20%)

A partir del archivo "iris.csv" que se corresponde con datos relacionados con tres tipos de plantas (el target es el tipo de planta):

- Mediante regresión logística, predice qué planta se corresponde con los ejemplos: [[15.7,2.8,9.5,0.1]] y [[1.7,3.8,4.5,11.3]]
- ¿Qué acierto tiene tu modelo? (score)
- Utiliza un conjunto de test con el 40% de los datos: ¿cambian los resultados anteriores? ¿por qué?
- Elimina la clase "Virginica" de tus datos. Repite el proceso anterior, ¿cambian los resultados?
- ¿Consideras la regresión logística un buen algoritmo para estos datos? ¿por qué?

In [234]:
import pandas as pd
import numpy as np

In [235]:
seed = 42

In [236]:
iris = pd.read_csv('../data/iris.csv')
iris

sepal.length  sepal.width  petal.length  petal.width    variety
0             5.1          3.5           1.4          0.2     Setosa
1             4.9          3.0           1.4          0.2     Setosa
2             4.7          3.2           1.3          0.2     Setosa
3             4.6          3.1           1.5          0.2     Setosa
4             5.0          3.6           1.4          0.2     Setosa
..            ...          ...           ...          ...        ...
145           6.7          3.0           5.2          2.3  Virginica
146           6.3          2.5           5.0          1.9  Virginica
147           6.5          3.0           5.2          2.0  Virginica
148           6.2          3.4           5.4          2.3  Virginica
149           5.9          3.0           5.1          1.8  Virginica

[150 rows x 5 columns]

In [237]:
X = np.array(iris.drop(['variety'], axis= 1))
y = np.array(iris['variety'])
print('X:', X.shape)
print('y:', y.shape)

X: (150, 4)
y: (150,)


## Inicializando modelo

In [238]:
from sklearn import linear_model

model = linear_model.LogisticRegression(max_iter=1000)

# TEST = 20%

In [239]:
from sklearn.model_selection import train_test_split

X_train_20, X_test_20, y_train_20, y_test_20 = train_test_split(X, y, test_size=0.2 ,random_state=seed)
print('X_train_20:', X_train_20.shape)
print('X_test_20:', X_test_20.shape)
print('y_train_20:', y_train_20.shape)
print('y_test_20:', y_test_20.shape)

X_train_20: (120, 4)
X_test_20: (30, 4)
y_train_20: (120,)
y_test_20: (30,)


## Cross Validation

In [240]:
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score

name_20 = 'Logistic Regression (TEST 20%)'

kfold = KFold(n_splits=8, random_state = seed, shuffle = True)
cv_results_20 = cross_val_score(model, X_train_20, y_train_20, cv = kfold, scoring = 'accuracy')
msg_20 = "%s: %f %s (%f)" % (name_20, cv_results_20.mean(), "+-", cv_results_20.std())
print(cv_results_20)
print('-----------------------------------------------------------')
print(msg_20)


[0.93333333 1.         1.         0.86666667 1.         0.86666667
 0.93333333 1.        ]
-----------------------------------------------------------
Logistic Refression (TEST 20%): 0.950000 +- (0.055277)


## La CV arroja buenos resultados. Ahora puedo entrenar a mi modelo con el conjunto de TRAIN completo

In [241]:
model.fit(X_train_20, y_train_20)

LogisticRegression(max_iter=1000)

## Predicción

In [242]:
model.predict([[15.7,2.8,9.5,0.1]])

# El modelo predice VIRGINICA

array(['Virginica'], dtype=object)

In [243]:
model.predict([[1.7,3.8,4.5,11.3]])

# El modelo predice nuevamente VIRGINICA

array(['Virginica'], dtype=object)

## Score del modelo

In [244]:
print('SCORES al entrenar el modelo con EL CONJUNTO TRAIN --> 20%')
print('\n')
print('Score TRAIN (TEST 20%):', model.score(X_train_20, y_train_20) * 100, '%')
print('Score TEST (TEST 20%):', model.score(X_test_20, y_test_20) * 100, '%')

SCORES al entrenar el modelo con EL CONJUNTO TRAIN --> 20%


Score TRAIN (TEST 20%): 97.5 %
Score TEST (TEST 20%): 100.0 %


## Si mi modelo tiene los resultados que deseo... ahora lo entreno con la totalidad de los datos

In [245]:
model.fit(X, y)
print('SCORES al entrenar el modelo con TODOS LOS DATOS')
print('\n')
print('Score (para todos los datos):', model.score(X, y) * 100, '%')
print('---------------------------------------------------------------------')
print('Score para conjunto TRAIN (TEST 20%):', model.score(X_train_20, y_train_20) * 100, '%')
print('---------------------------------------------------------------------')
print('Score TEST (TEST 20%):', model.score(X_test_20, y_test_20) * 100, '%')

SCORES al entrenar el modelo con TODOS LOS DATOS


Score (para todos los datos): 97.33333333333334 %
---------------------------------------------------------------------
Score para conjunto TRAIN (TEST 20%): 96.66666666666667 %
---------------------------------------------------------------------
Score TEST (TEST 20%): 100.0 %


# OBSERVACIÓN !!!!!!!!
### Conjunto TRAIN (TEST 20%) con modelo entrenado con ese mismo conjunto --> 97.5 %
### Conjunto TRAIN (TEST 20%) con modelo entrenado con todos los datos    --> 96.67 %

El conjunto train empeoró un poco al entrenar el modelo con la totalidad de los datos. ESTO DEPENDE SIEMPRE DE LA SEMILLA

## Predicción con el modelo entrenado con la totalidad de los datos

In [246]:
model.predict([[15.7,2.8,9.5,0.1]])

# El modelo predice VIRGINICA

array(['Virginica'], dtype=object)

In [247]:
model.predict([[1.7,3.8,4.5,11.3]])

# El modelo predice nuevamente VIRGINICA

array(['Virginica'], dtype=object)

## Matriz de confusión - TEST = 20%

In [248]:
from sklearn.metrics import confusion_matrix
prediccion_20 = model.predict(X_test_20)
print(confusion_matrix(y_test_20, prediccion_20))

[[10  0  0]
 [ 0  9  0]
 [ 0  0 11]]


---

# TEST = 40%

In [249]:
X_train_40, X_test_40, y_train_40, y_test_40 = train_test_split(X, y, test_size=0.4 ,random_state=seed)
print('X_train_40:', X_train_40.shape)
print('X_test_40:', X_test_40.shape)
print('y_train_40:', y_train_40.shape)
print('y_test_40:', y_test_40.shape)

X_train_40: (90, 4)
X_test_40: (60, 4)
y_train_40: (90,)
y_test_40: (60,)


## Cross Validation

In [250]:
name_40 = 'Logistic Refression - TEST 40%'

kfold = KFold(n_splits=8, random_state = seed, shuffle = True)
cv_results_40 = cross_val_score(model, X_train_40, y_train_40, cv = kfold, scoring = 'accuracy')
msg_40 = "%s: %f %s (%f)" % (name_40, cv_results_40.mean(), "+-", cv_results_40.std())
print(cv_results_40)
print('-----------------------------------------------------------')
print(msg_40)


[1.         0.91666667 1.         1.         0.81818182 1.
 0.81818182 0.90909091]
-----------------------------------------------------------
Logistic Refression - TEST 40%: 0.932765 +- (0.075134)


## La CV arroja buenos resultados. Ahora puedo entrenar a mi modelo con el conjunto de TRAIN completo

In [251]:
model.fit(X_train_40, y_train_40)

LogisticRegression(max_iter=1000)

## Predicción al entrenar el modelo con el conjunto TRAIN --> 20%

In [252]:
model.predict([[15.7,2.8,9.5,0.1]])

# El modelo predice VIRGINICA

array(['Virginica'], dtype=object)

In [253]:
model.predict([[1.7,3.8,4.5,11.3]])

# El modelo predice nuevamente VIRGINICA

array(['Virginica'], dtype=object)

## Score del modelo

In [254]:
print('SCORES al entrenar el modelo con EL CONJUNTO TRAIN --> 40%')
print('\n')
print('Score TRAIN (TEST 40%):', model.score(X_train_40, y_train_40) * 100, '%')
print('Score TEST (TEST 40%):', model.score(X_test_40, y_test_40) * 100, '%')

SCORES al entrenar el modelo con EL CONJUNTO TRAIN --> 40%


Score TRAIN (TEST 40%): 95.55555555555556 %
Score TEST (TEST 40%): 100.0 %


## Si mi modelo tiene los resultados que deseo... ahora lo entreno con la totalidad de los datos

In [255]:
model.fit(X, y)
print('SCORES al entrenar el modelo con TODOS LOS DATOS')
print('\n')
print('Score (Entrenado con la totalidad de los datos):', model.score(X_train_20, y_train_20) * 100, '%')
print('---------------------------------------------------------------------')
print('Score TRAIN (TEST 20%):', model.score(X_train_20, y_train_20) * 100, '%')
print('---------------------------------------------------------------------')
print('Score TEST (TEST 20%):', model.score(X_test_20, y_test_20) * 100, '%')

SCORES al entrenar el modelo con TODOS LOS DATOS


Score (Entrenado con la totalidad de los datos): 96.66666666666667 %
---------------------------------------------------------------------
Score TRAIN (TEST 20%): 96.66666666666667 %
---------------------------------------------------------------------
Score TEST (TEST 20%): 100.0 %


# OBSERVACIÓN !!!!!!!!
### Conjunto TRAIN (TEST 40%) con modelo entrenado con ese mismo conjunto --> 95.5 %
### Conjunto TRAIN (TEST 40%) con modelo entrenado con todos los datos    --> 96.67 %

El conjunto train mejoró un poco al entrenar el modelo con la totalidad de los datos. ESTO DEPENDE SIEMPRE DE LA SEMILLA

## Predicciones al entrenar el modelo CON TODOS LOS DATOS

In [256]:
print('Predicción 1:', model.predict([[15.7,2.8,9.5,0.1]]))
print('Predicción 2:', model.predict([[1.7,3.8,4.5,11.3]]))

Predicción 1: ['Virginica']
Predicción 2: ['Virginica']


## Matriz de confusión - TEST = 40%

In [257]:
prediccion_40 = model.predict(X_test_40)
print(confusion_matrix(y_test_40, prediccion_40))

[[23  0  0]
 [ 0 19  0]
 [ 0  0 18]]


# TEST 20% vs. TEST 40%
### No se produce un cambio significativo

---

## ELIMINANDO VIRGINICA

In [258]:
iris_vir = iris.drop(iris[iris['variety'] == 'Virginica'].index)
iris_vir

sepal.length  sepal.width  petal.length  petal.width     variety
0            5.1          3.5           1.4          0.2      Setosa
1            4.9          3.0           1.4          0.2      Setosa
2            4.7          3.2           1.3          0.2      Setosa
3            4.6          3.1           1.5          0.2      Setosa
4            5.0          3.6           1.4          0.2      Setosa
..           ...          ...           ...          ...         ...
95           5.7          3.0           4.2          1.2  Versicolor
96           5.7          2.9           4.2          1.3  Versicolor
97           6.2          2.9           4.3          1.3  Versicolor
98           5.1          2.5           3.0          1.1  Versicolor
99           5.7          2.8           4.1          1.3  Versicolor

[100 rows x 5 columns]

In [259]:
X_vir = np.array(iris_vir.drop(['variety'], axis= 1))
y_vir = np.array(iris_vir['variety'])
print('X_vir:', X_vir.shape)
print('y_vir:', y_vir.shape)


X_vir: (100, 4)
y_vir: (100,)


## Inicializando modelo

In [260]:
model_vir = linear_model.LogisticRegression(max_iter=1000)

# TEST 20% - Sin VIRGINICA

In [261]:
X_vir_train_20, X_vir_test_20, y_vir_train_20, y_vir_test_20 = train_test_split(X_vir, y_vir, test_size=0.2, random_state=seed)
print('X_vir_train_20:', X_vir_train_20.shape)
print('X_vir_test_20:', X_vir_test_20.shape)
print('y_vir_train_20:', y_vir_train_20.shape)
print('y_vir_test_20:', y_vir_test_20.shape)

X_vir_train_20: (80, 4)
X_vir_test_20: (20, 4)
y_vir_train_20: (80,)
y_vir_test_20: (20,)


## Cross validation

In [262]:
name_vir_20 = 'Logistic Regression - SIN VIRGINICA - TEST 20%'

kfold = KFold(n_splits=8, random_state=seed, shuffle=True)
cv_results_vir_20 = cross_val_score(model_vir, X_vir_train_20, y_vir_train_20, cv=kfold, scoring='accuracy')
msg_vir_20 = "%s: %f %s (%f)" % (name_vir_20, cv_results_vir_20.mean(), "+-", cv_results_vir_20.std())
print(cv_results_vir_20)
print('-----------------------------------------------------------')
print(msg_vir_20)

[1. 1. 1. 1. 1. 1. 1. 1.]
-----------------------------------------------------------
Logistic Regression - SIN VIRGINICA - TEST 20%: 1.000000 +- (0.000000)


## Los resultados son satisfactorios... entonces entreno mi modelo con el conjunto TRAIN

In [263]:
model_vir.fit(X_vir_train_20, y_vir_train_20)

LogisticRegression(max_iter=1000)

## Predicción al entrenar el modelo con TRAIN = 20% (SIN VIRGINICA)

In [264]:
print('Predicción 1:', model_vir.predict([[15.7,2.8,9.5,0.1]]))
print('Predicción 2:', model_vir.predict([[1.7,3.8,4.5,11.3]]))

# Tener en cuenta que ya no está VIRGINICA

Predicción 1: ['Versicolor']
Predicción 2: ['Versicolor']


In [265]:
print('Score TRAIN 20% Sin VIRGINICA:', model_vir.score(X_vir_train_20, y_vir_train_20))
print('Score TEST 20% Sin VIRGINICA:', model_vir.score(X_vir_test_20, y_vir_test_20))

Score TRAIN 20% Sin VIRGINICA: 1.0
Score TEST 20% Sin VIRGINICA: 1.0


## Los resultados son óptimos --> Entreno mi modelo con la totalidad de los datos

In [266]:
model_vir.fit(X_vir, y_vir)
print('Score (Entrenado con la totalidad de los datos) - SIN VIRGINICA:', model.score(X_train_20, y_train_20) * 100, '%')
print('-----------------------------------------------------------------------------------------')
print('Score TRAIN (TEST 20%) (Entrenado con la totalidad de los datos) - SIN VIRGINICA:', model.score(X_train_20, y_train_20) * 100, '%')
print('-----------------------------------------------------------------------------------------')
print('Score TEST (TEST 20%) (Entrenado con la totalidad de los datos) - SIN VIRGINICA:', model.score(X_test_20, y_test_20) * 100, '%')

Score (Entrenado con la totalidad de los datos) - SIN VIRGINICA: 96.66666666666667 %
-----------------------------------------------------------------------------------------
Score TRAIN (TEST 20%) (Entrenado con la totalidad de los datos) - SIN VIRGINICA: 96.66666666666667 %
-----------------------------------------------------------------------------------------
Score TEST (TEST 20%) (Entrenado con la totalidad de los datos) - SIN VIRGINICA: 100.0 %


## Predicción al entrenar el modelo con la TOTALIDAD DE LOS DATOS (SIN VIRGINICA)

In [267]:
print('Predicción 1:', model_vir.predict([[15.7,2.8,9.5,0.1]]))
print('Predicción 2:', model_vir.predict([[1.7,3.8,4.5,11.3]]))

# Tener en cuenta que ya no está VIRGINICA

Predicción 1: ['Versicolor']
Predicción 2: ['Versicolor']


## Matriz de confusión - TEST = 20% (Sin VIRGINICA)

In [268]:
prediccion_vir_20 = model_vir.predict(X_vir_test_20)
print(confusion_matrix(y_vir_test_20, prediccion_vir_20))

[[12  0]
 [ 0  8]]


---

# TEST = 40% - Sin VIRGINICA

In [269]:
X_vir_train_40, X_vir_test_40, y_vir_train_40, y_vir_test_40 = train_test_split(X_vir, y_vir, test_size=0.4, random_state=seed)
print('X_vir_train_40:', X_vir_train_40.shape)
print('X_vir_test_40:', X_vir_test_40.shape)
print('y_vir_train_40:', y_vir_train_40.shape)
print('y_vir_test_40:', y_vir_test_40.shape)

X_vir_train_40: (60, 4)
X_vir_test_40: (40, 4)
y_vir_train_40: (60,)
y_vir_test_40: (40,)


## Cross Validation

In [270]:
name_vir_40 = 'Logistic Regression - SIN VIRGINICA - TEST 40%'

kfold = KFold(n_splits=8, random_state=seed, shuffle=True)
cv_results_vir_40 = cross_val_score(model_vir, X_vir_train_40, y_vir_train_40, cv=kfold, scoring='accuracy')
msg_vir_40 = "%s: %f %s (%f)" % (name_vir_40, cv_results_vir_40.mean(), "+-", cv_results_vir_40.std())
print(cv_results_vir_40)
print('-----------------------------------------------------------')
print(msg_vir_40)

[1. 1. 1. 1. 1. 1. 1. 1.]
-----------------------------------------------------------
Logistic Regression - SIN VIRGINICA - TEST 40%: 1.000000 +- (0.000000)


## Los resultados son satisfactorios... entonces entreno mi modelo con el conjunto TRAIN


In [271]:
model_vir.fit(X_vir_train_40, y_vir_train_40)

LogisticRegression(max_iter=1000)

## Predicción al entrenar el modelo con TRAIN = 40% (SIN VIRGINICA)

In [272]:
print('Predicción 1:', model_vir.predict([[15.7,2.8,9.5,0.1]]))
print('Predicción 2:', model_vir.predict([[1.7,3.8,4.5,11.3]]))

# Tener en cuenta que ya no está VIRGINICA

Predicción 1: ['Versicolor']
Predicción 2: ['Versicolor']


In [273]:
print('Score TRAIN 40% Sin VIRGINICA:', model_vir.score(X_vir_train_40, y_vir_train_40))
print('Score TEST 40% Sin VIRGINICA:', model_vir.score(X_vir_test_40, y_vir_test_40))

Score TRAIN 40% Sin VIRGINICA: 1.0
Score TEST 40% Sin VIRGINICA: 1.0


## Los resultados son óptimos --> Entreno mi modelo con la totalidad de los datos

In [274]:
model_vir.fit(X_vir, y_vir)
print('Score (Entrenado con la totalidad de los datos) - SIN VIRGINICA:', model.score(X_vir, y_vir) * 100, '%')
print('-----------------------------------------------------------------------------------------')
print('Score TRAIN (TEST 40%) (Entrenado con la totalidad de los datos) - SIN VIRGINICA:', model.score(X_vir_train_40, y_vir_train_40) * 100, '%')
print('-----------------------------------------------------------------------------------------')
print('Score TEST (TEST 40%) (Entrenado con la totalidad de los datos) - SIN VIRGINICA:', model.score(X_vir_test_40, y_vir_test_40) * 100, '%')

Score (Entrenado con la totalidad de los datos) - SIN VIRGINICA: 97.0 %
-----------------------------------------------------------------------------------------
Score TRAIN (TEST 40%) (Entrenado con la totalidad de los datos) - SIN VIRGINICA: 100.0 %
-----------------------------------------------------------------------------------------
Score TEST (TEST 40%) (Entrenado con la totalidad de los datos) - SIN VIRGINICA: 92.5 %


## Predicción al entrenar el modelo con la TOTALIDAD DE LOS DATOS (SIN VIRGINICA)

In [275]:
print('Predicción 1:', model_vir.predict([[15.7,2.8,9.5,0.1]]))
print('Predicción 2:', model_vir.predict([[1.7,3.8,4.5,11.3]]))

# Tener en cuenta que ya no está VIRGINICA

Predicción 1: ['Versicolor']
Predicción 2: ['Versicolor']


## Matriz de confusión - TEST = 40% (Sin VIRGINICA)

In [276]:
prediccion_vir_40 = model_vir.predict(X_vir_test_40)
print(confusion_matrix(y_vir_test_40, prediccion_vir_40))

[[23  0]
 [ 0 17]]


## No se observan grandes cambios cuando se trabaja con la clase VIRGINICA y sin ella.

---

# Conclusión:

### Viendo los resultados obtenidos, se puede considerar que la Regresión Logística es un buen algoritmo para este caso.